include/polynomial.h

```diff
@@ -0,0 +1,112 @@
+#pragma once
+
+#include <cmath>
+#include <array>
+#include <vector>
+
+// Implemented according to https://en.wikipedia.org/wiki/Cubic_equation
+
+namespace quda
+{
+
+  inline std::vector<double> quadratic_formula(std::array<double, 3> coeff)
+  {
+
+    std::vector<double> z;
+    z.reserve(2);
+
+    double &a = coeff[0];
+    double &b = coeff[1];
+    double &c = coeff[2];
+
+    // a x^2 + b x + c = 0
+    if (a == 0) {
+      // actually a linear equation
+      if (b != 0) { z.push_back(-c / b); }
+    } else {
+      double delta = b * b - 4.0 * a * c;
+      if (delta >= 0) {
+        z.push_back((-b + std::sqrt(delta)) / (2.0 * a));
+        z.push_back((-b - std::sqrt(delta)) / (2.0 * a));
+      }
+    }
+
+    return z;
+  }
+
+  inline std::vector<double> cubic_formula(std::array<double, 4> coeff)
+  {
+
+    std::vector<double> t;
```

Comment:
`std::vector<double> t(3);`

Reply:
The size of the vector is used to indicate the number of real roots found, which is why I use `reserve` here.

Reply:
Sure, so why not declare the size of the vector in the constructor?

Reply:
Since the size is found later in the code.

Reply:
Ah, understood.  Thanks